sdk/helper/storagepacker/invalidate.go

```diff
@@ -0,0 +1,295 @@
+//
+// If we're the secondary, we shouldn't have to worry about concurrent access.
+func (s *StoragePackerV2) InvalidateItems(ctx context.Context, path string, newValue []byte) (present []*Item, deleted []*Item, err error) {
+	// Double-check that it's on our path, and figure out which bucket it should be
+	bucketKey, err := s.bucketKeyFromPath(path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var replacementBucket *LockedBucket
+	bucketRemoved := false
+
+	if newValue == nil {
+		bucketRemoved = true
+		// Create a placeholder so we can run the rest of the logic
+		// against an empty collection
+		replacementBucket = s.newEmptyBucket(bucketKey)
+	} else {
+		// Fake up a storage entry for the bucket and decode it
+		storage := &logical.StorageEntry{
+			Value: newValue,
+		}
+		replacementBucket, err = s.DecodeBucket(storage)
+		if err != nil {
+			return nil, nil, errwrap.Wrapf("unable to decode replicated bucket: {{err}}", err)
+		}
+		if replacementBucket.Key != bucketKey {
+			return nil, nil, fmt.Errorf("decoded bucket key %q doesn't match path component %q", replacementBucket.Key, bucketKey)
+		}
+	}
+
+	present = make([]*Item, 0)
+
+	originalBucket, err := s.GetBucket(ctx, bucketKey)
+	if err != nil {
+		return nil, nil, errwrap.Wrapf("problem finding original bucket: {{err}}", err)
+	}
+	originalBucket.Lock()
+	defer originalBucket.Unlock()
+
+	// Any item included in the new bucket should be reported present.
+	if !replacementBucket.HasShards && replacementBucket.ItemMap != nil {
+		for k, v := range replacementBucket.ItemMap {
+			present = append(present, &Item{
+				ID:   k,
+				Data: v,
+			})
+		}
+	}
+
+	// Is the new bucket a replacement, or new?  If it's new, nothing to compare against.
+	if originalBucket.Key == replacementBucket.Key {
+		// Any item not present in the new bucket *might* be deleted, or
+		// it might be elsewhere (in which case we want to report that value instead,
+		// to make sure the correct version is being used.)
+		maybeDeleted := itemDifferenceBetween(originalBucket, replacementBucket)
+		var revisit []*Item
+		deleted, revisit, err = s.identifyItemsAbsentOrShadowed(bucketKey, maybeDeleted)
+		if err != nil {
+			// Return partial information?
+			return present, nil, err
+		}
+		present = append(present, revisit...)
+	}
+
+	// Swap the replacement bucket in to the cache (or delete)
+	cacheKey := s.GetCacheKey(bucketKey)
+	s.bucketsCacheLock.RLock()
+	if bucketRemoved {
+		s.bucketsCache.Delete(cacheKey)
+	} else {
+		s.bucketsCache.Insert(cacheKey, replacementBucket)
+	}
+	s.bucketsCacheLock.RUnlock()
+
+	return present, deleted, nil
+
+}
+
+// Look up each item and determine whether another copy (or version) of it exists
+// in a different bucket.
+func (s *StoragePackerV2) identifyItemsAbsentOrShadowed(bucketKey string, maybeDeleted []*itemRequest) ([]*Item, []*Item, error) {
+	deleted := make([]*Item, 0)
+	revisit := make([]*Item, 0)
```

Comment:
These two fixed size arrays are appended to below but also have that same issue where they need things inserted at particular indexes instead, unless a different type of array is used.

Reply:
This also makes no sense - I needed my coffee. :-)